### PR TITLE
feat: environment image builds

### PR DESCRIPTION
**Pull request by ColeMurray**
## PR-10 · environment image builds

Phase-2 continuation of the multi-repo-sessions plan (follows #916). The environment prebuild pipeline (design §7.3/§7.4): environments get built as a unit — every repository cloned at its base branch, setup hooks run sequentially-fatally, filesystem snapshotted — and the resulting images are registered, invalidated, and reclaimed so PR-11 can select them at spawn. One migration: `0034` renames the `environment_images` provenance columns shipped in `0033` (PR-8) to repository nomenclature (`repositories_fingerprint`, `repository_shas`) — safe as a plain rename since nothing writes to the table until this PR lands.

Staged commits, as the plan sequences it: **Modal end-to-end first**, then the **Vercel/OpenComputer parity step**, followed by a review-fix pass and the member→repository nomenclature scrub.

### What it does

**Runtime (provider-neutral):** build mode now reports what only the build knows — per-repository clone provenance `repository_shas` (`[{repoOwner, repoName, baseSha}]`, one cross-language document shape used unchanged from runtime → callback → D1 → cron) and the baked `runtime_version` (read from the image's own `SANDBOX_VERSION`, so orchestrators never guess it from their code). Emitted via log events for Modal's orchestrator and via `report_success` for provider-session builds.

**Modal:** `build_environment_image` generalizes the repo-image worker (`SESSION_CONFIG.repositories` through the existing `create_build_sandbox` seam; redaction covers the clone token plus every build secret value; a completed build missing `repository_shas`/`runtime_version` fails with an actionable message instead of an opaque callback rejection). The 30-min cron gains an isolated **environments pass**: GET enabled environments + status → rebuild triggers, cheapest first — (1) no ready image matching the current repositories fingerprint, (3) runtime below the compatibility floor (or unparseable — fail closed), (2) per-repository `git ls-remote refs/heads/<branch>` drift (honors non-default branches, an improvement over repo images) — with a **per-tick trigger cap of 4** and per-environment concurrency 1.

**Control plane:** `EnvironmentImageStore` mirrors the repo-image state machine (conditional transitions, newer-ready races, `(created_at, id)` tiebreaks) with two model-owed differences: supersede scope is `(environment_id, provider)` — the fingerprint covers branches, so one live image per environment/provider — and rows are also superseded **out-of-band** (environment delete, secret change), so the cleanup pass gains a **superseded-artifact reaper** (artifact deleted first, row only after, failures retried next pass). The planner assembles build secrets as **global + environment** — the same set the environment's sessions get (§6.4 build/session parity) — through the PR-6 cap enforcement, and honors the primary repository's `buildTimeoutSeconds`. The workflow enforces **concurrency-1 at the trigger** (cron, save-hooks, and future "Rebuild now" all converge there) and **fails registration closed** on missing/unparseable `runtime_version` or `repository_shas` for *all* providers — an unversioned image must never pass PR-11's floor check. Routes: `/environment-images/{build-complete,build-failed,trigger/:id,status,enabled,mark-stale,cleanup}` mirroring the repo-image surface; `enabled` also serves each environment's fingerprint and the floor so the fingerprint algorithm never leaves the control plane.

**Save-hooks (§7.3/§7.4):** environment create/update triggers an immediate build (skipped when a ready image already matches the current repository set — trigger-1 evaluated eagerly); a secret change (PUT/DELETE/import) **supersedes every live image in the same request** — including in-flight builds, which are baking the outdated values — then kicks a rebuild. The supersede is awaited and fail-visible; the rebuild is detached and best-effort.

**Vercel/OpenComputer parity (commit 2):** both providers' hand-rolled branch-only build `SESSION_CONFIG`s are generalized so environment builds serialize the repositories-bearing payload; environment builds get the full provider-session lifecycle — bearer callback tokens (same token scheme as repo images), trigger-time session binding with teardown on failure, deferred snapshot/checkpoint finalization behind `waitUntil`, and build-sandbox cleanup hooks. Daytona untouched (base image), as designed.

### Tests
All green locally: control-plane unit 1656 (+44: workflow trigger/completion/failure/cleanup incl. provider-session paths, fingerprint/runtime-version parsing), control-plane integration 508 (+20 new `environment-images.test.ts`: store state machine vs real D1, cron-facing routes, callback auth + fail-closed registration matrix, secret-change supersede through the real PUT/DELETE routes, reaper leaves artifact-bearing rows for retry when the provider is unconfigured), sandbox-runtime 436 (+1 multi-repo repository_shas emission), modal-infra 195 (BuildLogResult + both cron passes); typecheck + eslint + prettier + ruff clean.

### Deliberate deviations from the plan doc
- **No `ttl=None` at snapshot time.** The Modal TTL change (#897) was rolled back and Modal is pinned at 1.4.3, where `snapshot_filesystem` has no `ttl` parameter and snapshots don't auto-expire — so **trigger 4 (provider-side expiry) is inert** and documented as the seam that activates with the snapshot-TTL follow-up. Nothing else in this PR depends on it.
- **`repositories_fingerprint` is computed control-plane-side at registration**, not echoed by the build callback. The same algorithm must run at spawn matching (PR-11); putting it in the callback would duplicate the hash across two languages. The callback carries only what the build alone knows: `repository_shas` + `runtime_version`.
- **`markRestoreFailed` is deferred to PR-11** with its caller (the spawn restore-failure fallback). It was specced here as "PR-0's twin", but PR-0 was reverted — repo-images has no such method today, and per the #909 precedent seams land with the code that uses them.
- **The concurrency-1 guard lives in the control-plane trigger path**, not cron-side like repo images — environment builds are triggered from three places (cron, save-hooks, later "Rebuild now"), so the guard sits where they converge. The cron still checks first to save the HTTP round-trip.
- **Mirror, not generalize.** The subsystem parallels repo-images (store/planner/workflow/adapters/errors) rather than parameterizing the merged, spawn-critical repo-image module; the genuinely shared leaves (callback-token scheme, provider policy, secrets cap enforcement, bearer-token extraction) are imported from their existing homes. Folding both into one image-build subsystem is a tracked follow-up refactor alongside the scoped-secrets-store unification.

### Not covered
- Live end-to-end (trigger → Modal build → callback → spawn selection) needs deployed infra — owed with the standing live-e2e pass; spawn selection itself is PR-11.
- Triggering a build via the route isn't integration-tested end-to-end (the harness has no Modal/Vercel credentials — the same suite-wide limitation as the SCM gap noted on #916); trigger orchestration is covered at the workflow layer with mocked adapters, and everything from the callback onward runs against real D1 through the real routes.

## Summary by CodeRabbit

* **New Features**
  * Introduced environment image builds across supported providers, including trigger, completion, failure, and cleanup workflows.
  * Added callback routes, status/enabled views, and maintenance endpoints for stale builds and cleanup.
  * Implemented multi-repository “repositories fingerprint” and environment-image build scheduling on save and after secrets changes.
* **Bug Fixes**
  * Enforced single in-flight builds and token-gated, fail-closed completion handling.
  * Improved superseding, late completion reporting, and reaping behavior (including artifact-retention edge cases).
* **Tests**
  * Added unit and integration coverage for fingerprints, workflow behavior, and scheduler rebuild decisions.

### CODE REVIEW

File: packages/control-plane/src/environment-images/workflow.ts [hunk 1]
```diff
@@ -0,0 +1,1096 @@
...
+      throw new EnvironmentImageWorkflowUnavailableError(
+        "Environment image provider is not configured"
+      );
+    }
+    if (!this.env.WORKER_URL) {
+      throw new EnvironmentImageWorkflowUnavailableError("WORKER_URL not configured");
+    }
+
+    const provider = this.provider;
+    const active = await this.store.getActiveBuild(environmentId, provider);
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/environment-images/workflow.ts:149]: Fixed in 62433de1. `registerBuild` is now the authoritative gate — a conditional `INSERT ... WHERE NOT EXISTS (... status = 'building')`, exactly as you suggested — and returns whether it inserted; a losing trigger reports the winner's buildId as `already_building` without planning or starting provider work. The `getActiveBuild` read stays only as a cheap short-circuit.

Covered by a workflow unit test and a real-D1 store test (which also proves an out-of-band supersede releases the slot). Worth noting the repo-images trigger path on main has the same read-before-write shape — fixing it there is part of the tracked repo/env image-subsystem unification follow-up.

File: packages/control-plane/src/db/environment-images.ts [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+      .bind(environmentId)
+      .all<EnvironmentImage>();
+
+    return result.results || [];
+  }
+
+  async getAllStatus(): Promise<EnvironmentImage[]> {
+    const result = await this.db
+      .prepare(
+        "SELECT * FROM environment_images WHERE status <> 'superseded' ORDER BY created_at DESC LIMIT 100"
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/db/environment-images.ts]: Fixed in 62433de1. The no-arg status endpoint now serves exactly the cron's question: ready + building rows joined to prebuild-enabled environments, uncapped. Unbounded is safe here because the state machine caps live rows per (environment, provider) at one ready + one building — a row cap would just reintroduce the starvation you describe, so it's gone entirely. Failed rows are excluded (the rebuild triggers never read them); the `?environment_id=` debug view keeps them.

The 100-row cap was inherited from repo-images' `getAllStatus`, which has the same latent issue — noted for the unification follow-up.

File: packages/modal-infra/src/scheduler/image_builder.py [hunk 3]
```diff
@@ -624,9 +975,20 @@ async def rebuild_repo_images():
     except Exception as e:
         log.error("scheduler.error", error=str(e))
 
+    # Environments pass — isolated so a repo-pass failure never starves it
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/modal-infra/src/scheduler/image_builder.py]: Fixed in 62433de1, taking the split you asked for: the repo section is now `_rebuild_repo_images_pass()`, where "no enabled repos" is a local `return 0`, and `rebuild_repo_images` is a thin orchestrator running the repo and environment passes in independent try/excepts.

One deliberate limit: the cron entrypoint keeps its historical name since renaming the Modal function would replace the deployed schedule; a rename can ride the image-subsystem unification follow-up. `test_skips_when_no_enabled_repos` now asserts the environment pass (enabled fetch + mark-stale + cleanup) still runs on an empty repo list.

File: packages/control-plane/src/routes/environment-images.ts [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+async function scheduleWorkflowTask(task: Promise<void>, ctx: RequestContext): Promise<void> {
+  if (ctx.executionCtx) {
+    ctx.executionCtx.waitUntil(task);
+    return;
+  }
+
+  await task;
+}
+
+async function parseCallbackBody<T>(request: Request): Promise<T | Response> {
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/routes/environment-images.ts:144]: Agreed the two route modules are drift-prone — this is the deliberate "mirror, not generalize" trade-off in the PR description: repo-images' routes are live spawn-critical surface, and extracting its callback/maintenance plumbing under this PR would couple the new feature to a refactor of merged code. The extraction is owned by the tracked "unify repo/env image-build subsystems" follow-up, and this round started the shared home: `parseMaxAgeMs` now lives in routes/shared.ts and both modules use it (62433de1). Happy to prioritize that follow-up right after the environments feature line lands.

File: packages/control-plane/src/sandbox/providers/vercel/provider.ts [hunk 5]
```diff
@@ -313,6 +340,83 @@ export class VercelSandboxProvider implements SandboxProvider {
     }
   }
 
+  /**
+   * Trigger a Vercel environment-image build (design §7.3). Same lifecycle as
+   * the repo-image build; the SESSION_CONFIG carries the member list so the
+   * list-native runtime clones and sets up every member.
+   */
+  async triggerEnvironmentImageBuild(
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/sandbox/providers/vercel/provider.ts:348]: Agreed on the end state — one provider-local build-sandbox helper that repo-image and environment-image triggers both call. This PR already unified the layer below it (`buildBuildEnvVars`/`buildBuildEnvironment` now take subject metadata + session config and serve both triggers); what's left duplicated is sandbox creation, tags, and the secret-store lifecycle. Folding that into a shared helper means reshaping the merged repo-image trigger too, so it's scoped to the tracked image-subsystem unification follow-up rather than this PR. The generalized env-var builders were shaped so that helper can land without another signature change.

File: packages/control-plane/src/environment-images/workflow.ts [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+ * Sequences planning, provider adapter calls, callback authorization, store
+ * state transitions, and best-effort artifact cleanup — the environment twin
+ * of RepoImageBuildWorkflow, trimmed to the provider_image callback mode
+ * Modal uses. HTTP parsing stays in routes, environment/secrets resolution in
+ * the planner, and provider API details in adapters.
+ *
+ * Public methods return successful domain outcomes and throw
+ * EnvironmentImageError subclasses for route-level error mapping.
+ */
+export class EnvironmentImageBuildWorkflow {
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/environment-images/workflow.ts:95]: Fair flag. Context that shaped the call: `repo-images/workflow.ts` is already 1036 lines on main, and the environment workflow deliberately mirrors its shape (trigger → callback → finalize → cleanup state machine) so the two stay twins until the tracked unification folds them into one subsystem. Decomposing only the new file now would diverge the shapes and make that unification harder, so I'd rather split both along the same seams (callback/finalization vs trigger/cleanup) in the follow-up. If you'd rather not merge at this size I can split the environment one first — but my preference is one decomposition, done once, across both.

File: packages/control-plane/src/environment-images/workflow.ts [hunk 7]
```diff
@@ -0,0 +1,1096 @@
...
+          break;
+        }
+      }
+
+      await this.cleanupCompletedBuild(provider, adapter, input, ctx);
+    } catch (e) {
+      if (adapter) {
+        await this.cleanupCompletedBuild(provider, adapter, input, ctx);
+      }
+      if (!finalized) {
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/environment-images/workflow.ts]: Fixed in 62433de1. `finalizeAndCommit` tracks whether the ready transition settled; when finalize succeeded but the transition threw, the catch best-effort-deletes the finalized snapshot and marks the build failed. Guarded on "commit never settled" rather than just `finalized` so a throw *after* a successful mark-ready (e.g. sandbox teardown) can't delete a live artifact or fail a ready row. Unit test covers the reclaim + markBuildFailed path.

File: packages/control-plane/src/routes/environment-images.ts [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+function buildCompleteCommand(
+  body: EnvironmentImageBuildCompleteBody
+): CompleteEnvironmentImageBuildCallback | Response {
+  const buildId = requireStringField(body.build_id, "build_id");
+  if (buildId instanceof Response) return buildId;
+
+  let buildDurationMs: number | undefined;
+  if (body.build_duration_seconds !== undefined) {
+    if (typeof body.build_duration_seconds !== "number") {
+      return error("build_duration_seconds is required", 400);
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/routes/environment-images.ts]: Fixed in 62433de1 — now "build_duration_seconds must be a number".

File: packages/control-plane/src/routes/environment-images.ts [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+  try {
+    body = (await request.json()) as typeof body;
+  } catch {
+    body = {};
+  }
+
+  const maxAgeMs =
+    body.max_age_seconds === undefined
+      ? DEFAULT_STALE_BUILD_MAX_AGE_MS
+      : body.max_age_seconds * MS_PER_SECOND;
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/routes/environment-images.ts]: Fixed in 62433de1. Both handlers parse through a shared `parseMaxAgeMs` (routes/shared.ts) that 400s on non-finite, negative, or non-number values and defaults only on truly-absent. Since repo-images' mark-stale/cleanup had the identical hole, they now go through the same helper. Integration test asserts null → 400 with no rows touched.

File: packages/modal-infra/tests/test_scheduler.py [hunk 10]
```diff
@@ -394,9 +493,19 @@ async def mock_post_side_effect(url, payload=None, **kwargs):
...
+
+        environment_stale_calls = [
+            c for c in mock_post.call_args_list if "environment-images/mark-stale" in str(c)
+        ]
+        assert len(environment_stale_calls) == 1
+
+        environment_cleanup_calls = [
+            c for c in mock_post.call_args_list if "environment-images/cleanup" in str(c)
+        ]
+        assert len(environment_cleanup_calls) == 1
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/modal-infra/tests/test_scheduler.py:540]: Added in 62433de1 alongside the scheduler split — `test_skips_when_no_enabled_repos` now mocks `_api_post` and asserts the environment pass runs its full skeleton while the repo pass stops at the empty list.

File: packages/control-plane/src/routes/shared.ts [hunk 11]
```diff
@@ -61,6 +61,34 @@ export function error(message: string, status = 400): Response {
...
+  if (body.max_age_seconds === undefined) return defaultMs;
+  if (
+    typeof body.max_age_seconds !== "number" ||
+    !Number.isFinite(body.max_age_seconds) ||
+    body.max_age_seconds < 0
+  ) {
+    return error("max_age_seconds must be a non-negative number", 400);
+  }
+  return body.max_age_seconds * 1000;
+}
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/routes/shared.ts:90]: @open-inspect[bot] please create a follow-up pull request resolving this
